django/utils/lru_cache.py

```diff
@@ -1,172 +1,5 @@
-                            link_next[PREV] = link_prev
-                            last = root[PREV]
-                            last[NEXT] = root[PREV] = link
-                            link[PREV] = last
-                            link[NEXT] = root
-                            stats[HITS] += 1
-                            return result
-                    result = user_function(*args, **kwds)
-                    with lock:
-                        root, = nonlocal_root
-                        if key in cache:
-                            # getting here means that this same key was added to the
-                            # cache while the lock was released.  since the link
-                            # update is already done, we need only return the
-                            # computed result and update the count of misses.
-                            pass
-                        elif _len(cache) >= maxsize:
-                            # use the old root to store the new key and result
-                            oldroot = root
-                            oldroot[KEY] = key
-                            oldroot[RESULT] = result
-                            # empty the oldest link and make it the new root
-                            root = nonlocal_root[0] = oldroot[NEXT]
-                            oldkey = root[KEY]
-                            oldvalue = root[RESULT]
-                            root[KEY] = root[RESULT] = None
-                            # now update the cache dictionary for the new links
-                            del cache[oldkey]
-                            cache[key] = oldroot
-                        else:
-                            # put result in a new link at the front of the list
-                            last = root[PREV]
-                            link = [last, root, key, result]
-                            last[NEXT] = root[PREV] = cache[key] = link
-                        stats[MISSES] += 1
-                    return result
-
-            def cache_info():
-                """Report cache statistics"""
-                with lock:
-                    return _CacheInfo(stats[HITS], stats[MISSES], maxsize, len(cache))
-
-            def cache_clear():
-                """Clear the cache and cache statistics"""
-                with lock:
-                    cache.clear()
-                    root = nonlocal_root[0]
-                    root[:] = [root, root, None, None]
-                    stats[:] = [0, 0]
-
-            wrapper.__wrapped__ = user_function
-            wrapper.cache_info = cache_info
-            wrapper.cache_clear = cache_clear
-            return update_wrapper(wrapper, user_function)
-
-        return decorating_function
+# Deprecate or remove this module when no supported version of Django still
+# supports Python 2. Until then, keep it to allow pluggable apps to support
+# Python 2 and Python 3 without raising a deprecation warning.
```

Comment:
This would require two changes: `import django.utils.lru_cache` => `try: import functools; except ImportError: import django.utils.lru_cache` => `import functools`.

By keeping the module around a little bit longer, third party apps can do only one change at virtually zero cost for us, so I think it's the right trade off there.

Credits for the suggestion go to Tim.